### DISCUSSION TITLE
implement hardware cursor to reduce input latency

### DISCUSSION THREAD

**Discussion by its0ka** [ppy/osu#22619] [Ideas]
almost any advanced player will say that osu with v-sync or windowed mode in windows is unplayable due to input lag. And some can't get rid of the lag because exclusive fullscreen mode doesn't work on amd and intel gpus. hw cursor will make it possible to play.
![IMG_20230212_123755](https://user-images.githubusercontent.com/125186204/218304840-3cff0460-37df-4cce-a583-9e6b1320f833.png)

**Comment by bdach** [comment]
what is "hardware cursor"? i have no idea what this would be. as far as i can tell that is not a thing that exists.

**Reply by its0ka** [reply]
https://stackoverflow.com/a/6958095
https://stackoverflow.com/a/21206890
windows cursor is a hardware cursor, osu cursor is a software cursor

**Reply by mk56-spn** [reply]
yeah uh, even if this DID give a noticeable reduction in lag id still be against it personally, since it cheeses mods such as no scope. ( im aware theres already ways to cheese it but this just makes the issue worse )

**Reply by bdach** [reply]
yeah i don't know about this. a lot of the claims about "input lag" at vsync are really a byproduct of stable's single-threaded nature. lazer is multi-threaded by default and tracks input at 1000 Hz so there should be a lot less of that.

i don't see this as necessary (and it would be a lot of platform-specific work).

**Reply by its0ka** [reply]
i used this program to show cursor in game https://dragonrisegames.com/yolomouse/about

**Reply by its0ka** [reply]
very sad that in order to play with good responsiveness I need to carefully choose a graphics card for my pc (amd has been ignoring broken exclusive fullscreen on their end for 6 months, idk about intel) or have a high refresh rate monitor to compensate broken exclusive fullscreen.

**Reply by mk56-spn** [reply]
Let's not be excessively dramatic now shall we . May I ask that you provide us with a video?

 The difference in cursor position in a still shot doesn't tell us much. That's a relatively large distance if you're moving your cursor slowly but it could be tiny if the cursor is moving really fast

**Reply by its0ka** [reply]
@mk56-spn https://drive.google.com/file/d/1Vy22gijJ7w2I_MvBAQkD7oO-Yt9U8iHY

**Reply by mk56-spn** [reply]
https://user-images.githubusercontent.com/74463310/218325092-5db50563-ced2-4dc3-8bd5-00732dd6a8c5.mp4

Compressed reupload of the provided video for anyone who doesn't want to have to download google drive files, it does seem the amount of delay is non negligible.

**Reply by Cootz** [reply]
Looks kind of bad. I also feel like `osu!lazer` have *heavier* mouse input compare to the stable one

**Comment by peppy** [comment]
I've been waiting my whole life for someone to suggest this.

It is on my mind as something we may offer in the future, but I haven't prioritised it until now because:

- Cursor trail will be very hard to make look good (it will be disconnected if the user is running vsync, for instance)
- Generally we hope that using the multi-threaded architecture of lazer, we can already provide similar low latency cursor. As you can see from the [video above](https://github.com/ppy/osu/discussions/22619#discussioncomment-4950198) it's on-par with most common settings. I have already tested this myself over the last few years which is a reason I have not investigated further. In fact in some testing setups lazer's cursor was *lower latency* than the windows cursor (go figure, I'm not sure why).
- While animating the hardware cursor is possible, it will be very hard to get right (ie. offer the same customisation that is exposed to classic skins). We would have to rasterise the cursor and probably do so ahead of time.
- If the game is running slow enough for the hardware cursor to be hugely beneficial (ie. vsync), it probably also going to cause audio latency of equal proportions. Therefore our focus is on getting the game to perform well at 120-480fps update so we can ensure all aspects of the game are low latency, not just the cursor.
- Hardware cursor is not supported on all operating systems.

Hopefully this gives some insight into the issue. I'm not going to say that we won't ever support it, but I'd like to hope that we don't *need* to because what we have already guarantees low latency cursor rendering. Most games which support hardware cursor have locked render loops at 60fps or so, and triple buffering / render-to-texture effects applied, where it becomes much more beneficial.
